Progress bar

borjainglefernandez opened the pull request:
## Status
- Done.

## GitHub Issues addressed (`#IssueNumber`)
- This PR closes #296 

## What I did
- Added progress bar
- Added message for the progress bar
- Added functionality to both the message and progress bar itself so that they update with each word typed

## Screenshots
- Before
![image](https://user-images.githubusercontent.com/65180311/108611418-67ebae00-73ac-11eb-8b83-3c7b552613cb.png)

- After
![image](https://user-images.githubusercontent.com/65180311/108611402-51dded80-73ac-11eb-9c6b-73582314cbbd.png)
![image](https://user-images.githubusercontent.com/65180311/108611406-54404780-73ac-11eb-950a-3a63c5dfd9f5.png)
![image](https://user-images.githubusercontent.com/65180311/108611407-56a2a180-73ac-11eb-89db-826060ddb92a.png)
![image](https://user-images.githubusercontent.com/65180311/108611408-586c6500-73ac-11eb-893b-6a2bcfabfbb5.png)

## Tests
- A brief explanation of tests done/written or how reviewers can test your work
-- Go to the review page and start typing and make sure that the progress bar incrementally begins moving. Also, ensure that the bar doesn't move based on spaces. 
--Test to see if the thresholds are correct (i.e. after 200 words the message changes and the bar turns from orange to yellow)

## Questions/Discussions/Notes
- Front end design could use some work?
- I lazily came up with messages... definitely could improve the messages...

## Merging the PR
- Who should merge this PR?
  - [ ] I will merge it myself
  - [ ] The last reviewer to approve can merge it
- Should the commit history be preserved in the base branch, or is it okay to combine all of them into a single commit ("squash-merge")?
  - [ ] preserve the history
  - [ ] squash-merge

j-alicia-long reviewed (commented):
Great job! Recommend reaching out to design for UI feedback if you haven't already.

Some suggestions I'd have are:
- Put the bar under the review label "Tell us more", or possibly below the input box itself.
- Make it a lil thinner and less bold in color so it's not super distracting. Consider some pastels or maybe a orange-to-blue to match our theme? shrug

borjainglefernandez commented:
> Great job! Recommend reaching out to design for UI feedback if you haven't already.
> 
> Some suggestions I'd have are:
> 
> * Put the bar under the review label "Tell us more", or possibly below the input box itself.
> * Make it a lil thinner and less bold in color so it's not super distracting. Consider some pastels or maybe a orange-to-blue to match our theme? shrug

Those are all great ideas! I'll reach out to the design team and implement your suggestions.

n4mahajan commented:
Similar to what Jenn said, I would recommend putting the progress bar under the input box that the user writes in as I feel this is most consistent with where progress bars go? And perhaps the bar should also be centered with the page? Perhaps worth at least trying out. I think it might be a little clunky putting it between the "Tell us more" label and input box. Love the dynamic progress bar though, I think it definitely adds to the UX. 

I think the color scheme of the progress bar could still use some work though and I'm not sure it adequately symbolizes the user's progress. While I agree with Jenn that the colors before may have been too bold, I feel like they might be too passive now but I think that's definitely up for debate. I think particularly the yellow portion of the progress bar could use an update.

n4mahajan commented on tcf_website/static/reviews/new_review.js:
```diff
@@ -124,8 +124,48 @@ jQuery(function($) {
+        // Chose to count words based on how many spaces there are
+        if (s === " ") {
+            count++;
+        }
```
From what I can tell by trying it and the code, this method seems to include punctuation and really any text that might not be a word. This quite clearly wouldn't reflect the actual word count, but the hassle of trying to count words **exclusively** may not be worth the hassle of a rather trivial thing since most of what a user is writing is indeed words.

borjainglefernandez replied:
hmmmm good catch. Maybe it's worth having a list of all the characters that would count as contributing to words and only count those?

n4mahajan replied:
From what I can tell, you seemed to make it so that only actual words are counted as opposed to other special characters, so awesome job again! I'm not sure if there's any more efficient method out there to do it but the speed/performance doesn't seem hindered.

n4mahajan commented on tcf_website/static/reviews/new_review.js:
```diff
@@ -124,8 +124,48 @@ jQuery(function($) {
+
+        // Set the width of the bar to the what percent of 300 words the current review is
+        $("#reviewprogressbar").width(1000 * (numberOfWords / 300));
+        if (numberOfWords < 100) {
```
When I was testing the progress bar, it seemed the color of the bar was the same both when the word count was less than 100 and between 100 and 200. I feel like it should change colors once the user passes 100 words as the text does so that their progress in writing the review is further enforced?

borjainglefernandez replied:
It does change colors from red to orange but I think you just caught a bug where if you delete everything, the color stays the same

n4mahajan replied:
Yeah I see from the code it does seem to change colors, but I think the coloring was so light the change was practically unnoticeable? Might be worth making more distinct if possible. It looks like you updated the colors now and the change is definitely noticeable now which is good. Looks like you fixed the bug where it didn't update immediately after deleting (vice versa, it also updates immediately when you undo the deletion) so great job! I love how dynamic it is.

n4mahajan commented:
> Similar to what Jenn said, I would recommend putting the progress bar under the input box that the user writes in as I feel this is most consistent with where progress bars go? And perhaps the bar should also be centered with the page? Perhaps worth at least trying out. I think it might be a little clunky putting it between the "Tell us more" label and input box. Love the dynamic progress bar though, I think it definitely adds to the UX.
> 
> I think the color scheme of the progress bar could still use some work though and I'm not sure it adequately symbolizes the user's progress. While I agree with Jenn that the colors before may have been too bold, I feel like they might be too passive now but I think that's definitely up for debate. I think particularly the yellow portion of the progress bar could use an update.

n4mahajan commented:
Also would add that the progress bar only updates when the user presses space (i.e. doesn't update until you press space even if you delete a bunch of words). I don't know how feasible it is, but may be worth looking into making the progress bar even more dynamic so that it responds even more to the user's actions.

n4mahajan commented:
> Also would add that the progress bar only updates when the user presses space (i.e. doesn't update until you press space even if you delete a bunch of words). I don't know how feasible it is, but may be worth looking into making the progress bar even more dynamic so that it responds even more to the user's actions.

Ah just kidding, sorry, it does update when you delete words but you have to get rid of the space in between first. Leaving this here to avoid confusion. Although, if I were to highlight a bunch of words and press delete, it doesn't automatically update, so I think the general idea of making the progress bar even more dynamic still stands. Again though, might not be worth the trouble because it works well as is.

borjainglefernandez commented:
> Similar to what Jenn said, I would recommend putting the progress bar under the input box that the user writes in as I feel this is most consistent with where progress bars go? And perhaps the bar should also be centered with the page? Perhaps worth at least trying out. I think it might be a little clunky putting it between the "Tell us more" label and input box. Love the dynamic progress bar though, I think it definitely adds to the UX.
> 
> I think the color scheme of the progress bar could still use some work though and I'm not sure it adequately symbolizes the user's progress. While I agree with Jenn that the colors before may have been too bold, I feel like they might be too passive now but I think that's definitely up for debate. I think particularly the yellow portion of the progress bar could use an update.

I agree with the yellow, maybe we could just make that blue? I'm not sure what other color we could do though. I'll move the progress bar below but I am not sure what you mean by centered with the page? As it is right now it is centered with the review form.

borjainglefernandez commented:
I just fixed the bugs that were present and moved where the progress bar was. As far as the colors, I decided to leave them as they are now just until we figure out what's best.

n4mahajan commented:
> > Similar to what Jenn said, I would recommend putting the progress bar under the input box that the user writes in as I feel this is most consistent with where progress bars go? And perhaps the bar should also be centered with the page? Perhaps worth at least trying out. I think it might be a little clunky putting it between the "Tell us more" label and input box. Love the dynamic progress bar though, I think it definitely adds to the UX.
> > I think the color scheme of the progress bar could still use some work though and I'm not sure it adequately symbolizes the user's progress. While I agree with Jenn that the colors before may have been too bold, I feel like they might be too passive now but I think that's definitely up for debate. I think particularly the yellow portion of the progress bar could use an update.
> 
> I agree with the yellow, maybe we could just make that blue? I'm not sure what other color we could do though. I'll move the progress bar below but I am not sure what you mean by centered with the page? As it is right now it is centered with the review form.

Yeah, would be good to get some input from the design team but I think the color scheme is pretty solid now. I think maybe the main thing to look at would be to make the text with the progress bar (i.e. "Thank you for your in depth review. The tCF team and other users appreciate your effort!") more interactive/appealing if possible? Again a lot of the stuff I've pointed out is rather nitpicky but may be good for UI/UX.

borjainglefernandez commented:
> . I think maybe the main thing to look at would be to make the text with the progress bar (i.e. "Thank you for your in depth review. The tCF team and other users appreciate your effort!"

Any suggestions about making the test more interactive? The messages themselves def can be improved but I don't see how it can be more interactive.

borjainglefernandez commented:
> > . I think maybe the main thing to look at would be to make the text with the progress bar (i.e. "Thank you for your in depth review. The tCF team and other users appreciate your effort!"
> 
> Any suggestions about making the test more interactive? The messages themselves def can be improved but I don't see how it can be more interactive.

I thought about it more and one thing we can do is show the word count explicitly? And maybe allow the user to customize what kind of colors they want (e.g. Stop Light (red,yellow, green), tCF (orange to blue), Rainbow(ROYGBIV)). Like have a preset color scheme for each option but allow the user to decide.

n4mahajan commented:
> > > . I think maybe the main thing to look at would be to make the text with the progress bar (i.e. "Thank you for your in depth review. The tCF team and other users appreciate your effort!"
> > 
> > 
> > Any suggestions about making the test more interactive? The messages themselves def can be improved but I don't see how it can be more interactive.
> 
> I thought about it more and one thing we can do is show the word count explicitly? And maybe allow the user to customize what kind of colors they want (e.g. Stop Light (red,yellow, green), tCF (orange to blue), Rainbow(ROYGBIV)). Like have a preset color scheme for each option but allow the user to decide.

I like the idea of an active word count; could look pretty cool tbh. The color customization might be harder tho? Especially if you have to store the preference that each user has if they change the color scheme, but could help enforce our branding to some extent too. Could maybe even make the progress bar "pulsate" or "flash" if that makes sense although honestly anything beyond what you have now in terms of additional features may not be worth your time.

j-alicia-long commented on tcf_website/templates/reviews/review_form_content.html:
```diff
@@ -39,9 +39,17 @@ <h2 class="h4 review-form-header form-group col-12 mt-4 mb-2">
+    <p class="medium" id="progressbarmessage"></p>
+  </div>
+        <!-- Progress Bar -->
+  <div class="progress" style="height: 25px;">
```
We usually try to keep styling in the css files!

btw I think it'd look nicer if the progress bar was a little thinner (15px?) and there was less spacing between the prompt text and the bar.

borjainglefernandez replied:
Yes, thanks for pointing that out it's now 15 px.

j-alicia-long reviewed (changes requested):
Just checked it out locally and found a bug: the progress bar isn't responsive for different window sizes (see mobile below)

![image](https://user-images.githubusercontent.com/4724192/110548964-5cfc8180-80ff-11eb-9307-241723e423ee.png)

borjainglefernandez commented:
![image](https://user-images.githubusercontent.com/65180311/110576710-32c2b800-812f-11eb-86a6-262aec27f7dd.png)

![image](https://user-images.githubusercontent.com/65180311/110576684-2a6a7d00-812f-11eb-9500-4aa8b9b894ba.png)


![image](https://user-images.githubusercontent.com/65180311/110576671-25a5c900-812f-11eb-83eb-991ef3c97b25.png)

![image](https://user-images.githubusercontent.com/65180311/110576650-1de62480-812f-11eb-9c15-3ef663add289.png)

This is what it looks like now.

j-alicia-long reviewed (approved):
Small bug remaining: Progress bar fill doesn't change when you resize the browser?

![image](https://user-images.githubusercontent.com/4724192/110685297-1fa1fd80-81ac-11eb-9752-e6de95640f53.png)

Since this is an edge case (who's going to resize in the middle of writing a review) I'll just approve it and make a new issue for it
